codechain/crypto/ecdsa.py

```diff
@@ -0,0 +1,123 @@
+
+    privkey = PrivateKey(priv)
+    sig_check = privkey.sign_recoverable(message, hasher=None)
+
+    return sig_check
+
+
+# Checks if the signature from signEcdsa is correct.
+
+
+def verify_ecdsa(message, signature, pub):
+    if not isinstance(message, (bytes, bytearray)) or len(message) != 32:
+        raise ValueError(f"Invalid message: {message}")
+    if not isinstance(signature, (bytes, bytearray)) or len(signature) != 65:
+        raise ValueError(f"Invalid signature: {signature}")
+    if not isinstance(pub, (bytes, bytearray)) or len(pub) != 64:
+        raise ValueError(f"Invalid public key: {pub}")
+
+    pubkey = PublicKey(b"\x04" + pub)
+
+    r = big_endian_to_int(signature[0:32])
+    s = big_endian_to_int(signature[32:64])
+
+    low_s = coerce_low_s(s)
+    der_encoded_signature = bytes(two_int_sequence_encoder(r, low_s))
+
+    return verify_signature(
+        der_encoded_signature, message, pubkey.format(compressed=False), hasher=None
+    )
+
+
+# Gets public key from the message and ECDSA signature.
+
+
+def recover_ecdsa(message, signature):
+    if not isinstance(message, (bytes, bytearray)) or len(message) != 32:
+        raise ValueError(f"Invalid message: {message}")
+    if not isinstance(signature, (bytes, bytearray)) or len(signature) != 65:
+        raise ValueError(f"Invalid private key: {signature}")
+
+    return PublicKey.from_signature_and_message(signature, message, hasher=None).format(
+        compressed=False
+    )[1:]
+
+
+def int_to_big_endian(value: int) -> bytes:
+    return value.to_bytes((value.bit_length() + 7) // 8 or 1, "big")
+
+
+def big_endian_to_int(value: bytes) -> int:
+    return int.from_bytes(value, "big")
+
+
+def coerce_low_s(value: int) -> int:
+    """Coerce the s component of an ECDSA signature into its low-s form.
+
+    See https://bitcoin.stackexchange.com/questions/83408/in-ecdsa-why-is-r-%E2%88%92s-mod-n-complementary-to-r-s  # noqa: W501
+    or https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2.md.
+    """
+    return min(value, -value % SECPK1_N)
+
+
+def _encode_int(primitive: int):
+    # See: https://docs.microsoft.com/en-us/windows/desktop/seccertenroll/about-integer
+
+    # Integer tag
+    yield 0x02
+
+    encoded = int_to_big_endian(primitive)
+    if encoded[0] >= 128:
+        # Indicate that integer is positive
+        yield len(encoded) + 1
+        yield 0x00
+    else:
+        yield len(encoded)
+
+    yield from encoded
+
+
+def two_int_sequence_encoder(signature_r: int, signature_s: int):
+    """
+    Encode two integers using DER, defined as:
+
+    ::
+
+        ECDSASpec DEFINITIONS ::= BEGIN
+              ECDSASignature ::= SEQUENCE {
+                 r   INTEGER,
+                 s   INTEGER
+             }
+        END
+
+    Only a subset of integers are supported: positive, 32-byte ints.
+
+    See: https://docs.microsoft.com/en-us/windows/desktop/seccertenroll/about-sequence
+    """
+    # Sequence tag
+    yield 0x30
```

Comment:
Why do we encode asn.1 manually? Why don't just use https://pypi.org/project/asn1?

Reply:
The lib is not fully documented so that I decided to use my own code.

Reply:
Also, eth-keys is using this code, not the asn1 package.